source/common/router/retry_state_impl.cc

```diff
@@ -374,41 +412,82 @@ bool RetryStateImpl::wouldRetryFromHeaders(const Http::ResponseHeaderMap& respon
            (retry_on_ & RetryPolicy::RETRY_ON_GRPC_UNAVAILABLE)) ||
           (status.value() == Grpc::Status::Internal &&
            (retry_on_ & RetryPolicy::RETRY_ON_GRPC_INTERNAL))) {
-        return true;
+        return RetryDecision::RetryWithBackoff;
       }
     }
   }
 
-  return false;
+  return RetryDecision::NoRetry;
 }
 
-bool RetryStateImpl::wouldRetryFromReset(const Http::StreamResetReason reset_reason) {
+RetryState::RetryDecision
+RetryStateImpl::wouldRetryFromReset(const Http::StreamResetReason reset_reason,
+                                    absl::optional<bool> was_using_alt_svc, bool& disable_alt_svc) {
+  ASSERT(!disable_alt_svc);
   // First check "never retry" conditions so we can short circuit (we never
   // retry if the reset reason is overflow).
   if (reset_reason == Http::StreamResetReason::Overflow) {
-    return false;
+    return RetryDecision::NoRetry;
+  }
+
+  const bool consider_disable_alt_svc = Runtime::runtimeFeatureEnabled(
+      "envoy.reloadable_features.conn_pool_new_stream_with_early_data_and_alt_svc");
+  if (consider_disable_alt_svc) {
+    if (reset_reason != Http::StreamResetReason::ConnectionFailure) {
+      ASSERT(was_using_alt_svc.has_value());
+      if (*was_using_alt_svc) {
+        // Retry any post-handshake failure immediately with no alt_svc if the request was sent over
+        // Http/3.
+        disable_alt_svc = true;
+        // TODO(danzh) consider making the retry configurable.
+        return RetryDecision::RetryNoBackoff;
+      }
+    } else {
+      if (was_using_alt_svc.has_value()) {
```

Comment:
This is a bit confusing to me. What's the difference between was_using_alt_svc = false and was_using_alt_svc.has_value() = false? In general, optional<bool> can be a bit confusing. I wonder if we should consider using a 3-state enum for this?